pybb/models/mixins.py

```diff
@@ -0,0 +1,89 @@
+from itertools import chain
+from django.contrib.postgres.fields import ArrayField
+from django.db import models, transaction
+from django.db.models import signals
+
+from base import ModelBase
+
+
+class ParentForumQuerysetMixin(object):
+    def has_parent(self, parent):
+        return self.filter(parents__contains=[parent.id])
+
+
+class ParentForumManagerMixin(object):
+    def get_queryset(self):
+        return ParentForumQuerysetMixin(self.model)
+
+    def has_parent(self, *args, **kwargs):
+        return self.get_queryset().has_parent(*args, **kwargs)
+
+    def contribute_to_class(self, cls, name):
+        signals.pre_save.connect(self.parents_pre_save, sender=cls)
+        signals.post_save.connect(self.parents_post_save, sender=cls)
+        return super(ParentForumManagerMixin, self).contribute_to_class(cls, name)
+
+    def parents_pre_save(self, sender, instance, **kwargs):
+        if ((instance.parents and instance.parents[0] == instance.forum_id) or
+                (not instance.parents and not instance.forum_id)):
+            instance._has_moved = False
+            return
+        parent = instance.forum
+        parents = []
+        while parent:
+            parents.append(parent.id)
+            parent = parent.forum
+
+        instance.parents = parents
+        instance._has_moved = True
+
+    @transaction.atomic
+    def parents_post_save(self, sender, instance, **kwargs):
+        from pybb.models import Forum, Topic
+
+        if not (sender == Forum and instance._has_moved):
```

Comment:
Not possible, the mixin can be used on any manager (i our case, ForumManager and TopicManager), so self.model is not guaranteed to be Forum